tracing-subscriber/src/fmt/format/json.rs

```diff
@@ -217,23 +217,28 @@ where
                 None
             };
 
-            if self.format.display_span_list && current_span.is_some() {
-                serializer.serialize_entry(
-                    "spans",
-                    &SerializableContext(&ctx.ctx, format_field_marker),
-                )?;
+            if !self.format.flatten_event {
+                use tracing_serde::fields::AsMap;
+                serializer.serialize_entry("fields", &event.field_map())?;
+            }
```

Comment:
If we are not flattening the event's fields into the root JSON object, it looks like we will now emit the fields first, llike this:
```json
{
    "timestamp":"Feb 20 11:28:15.096",
    "level":"INFO",
    "fields":{"message":"some message","key":"value"}
    "target":"mycrate",
    "span":{name":"leaf"},
    "spans":[{"name":"root"},{"name":"leaf"}],
}
```
But if we _are_ flattening the event fields, it looks like the fields are still last, so it will look like this:
```json
{
    "timestamp":"Feb 20 11:28:15.096",
    "level":"INFO",
    "target":"mycrate",
    "span":{name":"leaf"},
    "spans":[{"name":"root"},{"name":"leaf"}],
    "message":"some message",
    "key":"value"
}
``

Do you think we should maybe move the code from the `else` here https://github.com/tokio-rs/tracing/blob/be827b12acb9e2626c63bce31242123834e3cf0e/tracing-subscriber/src/fmt/format/json.rs#L266-L271 to up here, as well, so that the ordering is consistent regardless of flattening?

Reply:
Sure! I had to introduce a new function to `SerdeMapVisitor` to return the underlying serializer back to the caller to make this work – hopefully that's not a problem!

---

I'm not sure if the JSON standard specifies any particular action when a key appears twice. Both Firefox and Chrome seem to use the last specified value:
* `JSON.parse('{ "target": "mycrate", "target": "practice" }')` will return `{ "target": "practice" }`
* `JSON.parse('{ "target": "practice", "target": "mycrate" }')` will return `{ "target": "mycrate" }` 

So this *could* be considered a breaking change in the pathological case where a consumer with flattening on emits a key that is one of the known keys.

```rust
event!(Level::INFO, archer = "Bob", bow = "yew", arrow = "dx-487", target = "practice", hit = "bullseye");
```

Reply:
> Sure! I had to introduce a new function to `SerdeMapVisitor` to return the underlying serializer back to the caller to make this work – hopefully that's not a problem!

Okay, that's fine, we'll just have to make sure to publish a new `tracing-serde` release before we can publish a new `tracing-subscriber`. Not a problem, just making a note of this.



> I'm not sure if the JSON standard specifies any particular action when a key appears twice. Both Firefox and Chrome seem to use the last specified value:
> 
>     * `JSON.parse('{ "target": "mycrate", "target": "practice" }')` will return `{ "target": "practice" }`
> 
>     * `JSON.parse('{ "target": "practice", "target": "mycrate" }')` will return `{ "target": "mycrate" }`
> 
> 
> So this _could_ be considered a breaking change in the pathological case where a consumer with flattening on emits a key that is one of the known keys.

Hmm. That _could_ be an issue. The names "target" and "file", which appear in the metadata, are both potentially common enough that I think potential clashes with a field name might be worth worrying about. We might want to make other changes (out of scope for this PR) to handle this when the fields are being flattened...maybe prefixing the metadata fields with "metadata_" or something? Of course, that would break the JSON format as well, so it would have to wait for 0.3 before we can address it.